Relocate the hosts file to /usr/share/ceph-ansible

TimothyAsirJeyasing opened the pull request:
BZ#1879836
https://github.com/red-hat-storage/cockpit-ceph-installer/issues/79
Signed-off-by: Timothy Asir Jeyasingh <tjeyasin@redhat.com>

TimothyAsirJeyasing committed:
Relocate the hosts file to /usr/share/ceph-ansible

create /usr/share/ceph-ansible/hosts if it doesn't exist
link /usr/share/ansible-runner-service/inventory/hosts to ceph-ansible-hosts

Bug: BZ#1879836
https://github.com/red-hat-storage/cockpit-ceph-installer/issues/79
Signed-off-by: Timothy Asir Jeyasingh <tjeyasin@redhat.com>

TimothyAsirJeyasing commented:
@pcuzner Please review

pcuzner commented on utils/ansible-runner-service.sh:
```diff
@@ -414,17 +414,20 @@ check_access() {
 manage_ansible_hosts_file() {
-    echo "Linking the default ansible hosts file to the runner-service inventory."
+    echo "Linking the runner service inventory to ceph-ansible hosts"
+    ln -sf /usr/share/ceph-ansible/hosts /usr/share/ansible-runner-service/inventory/hosts
```
this hosts file may not exist - so you may need to check that it's there and if not touch it

pcuzner commented on utils/ansible-runner-service.sh:
```diff
@@ -414,17 +414,20 @@ check_access() {
+    ln -sf /usr/share/ceph-ansible/hosts /usr/share/ansible-runner-service/inventory/hosts
+
+    echo "Linking the default ansible hosts file to the ceph-ansible hosts file."
     if [ -h "$DEFAULT_ANSIBLE_HOSTS" ]; then
```
we probably need to reassess this. we were linking to the default ansible hosts file - now we're not, so the logic here needs review.

We should be 
creating /usr/share/ceph-ansible/hosts - if it doesn't exist
linking /usr/share/ansible-runner-service/inventory/hosts to /usr/share/ceph-ansible-hosts
checking to see if the /usr/share/ceph-ansible/hosts contains hosts - if so, we should move it to one side, tell the admin and set up a new one to use with the installer. (remember ansible supports yaml, and ini format inventory files - so we need to get the inventory to a known state here)

we should no longer be manipulating /etc/ansible/hosts - wither directly or via symlink

pcuzner reviewed (changes requested):
Needs a bit more thought about how we handle the hosts file I think

pcuzner commented on utils/ansible-runner-service.sh:
```diff
@@ -415,36 +415,29 @@ check_access() {
+    [ ! -f $CEPH_ANSIBLE_HOSTS ] && touch $CEPH_ANSIBLE_HOSTS
+
+    # Set up a new hosts file if the hosts file contains hosts details already or it is an INI file
+    if grep -q -e 'hosts:' -e ']' "$CEPH_ANSIBLE_HOSTS" 2>/dev/null; then
```
an inventory could just be a list of servers (not grouped), so perhaps just a -s test would suffice i.e. it exists and it has content.

pcuzner commented on utils/ansible-runner-service.sh:
```diff
@@ -416,7 +416,7 @@ check_access() {
     echo "Linking the runner service inventory to ceph-ansible hosts"
     # create hosts file if it doesn't exist
-    [ ! -f $CEPH_ANSIBLE_HOSTS ] && touch $CEPH_ANSIBLE_HOSTS
+    [ ! -s $CEPH_ANSIBLE_HOSTS ] && touch $CEPH_ANSIBLE_HOSTS
```
this isn't what I was talking about. this should just be file exists (-f) and please use the variable in quotes, then see below

pcuzner commented on utils/ansible-runner-service.sh:
```diff
@@ -416,7 +416,7 @@ check_access() {
+    [ ! -s $CEPH_ANSIBLE_HOSTS ] && touch $CEPH_ANSIBLE_HOSTS
 
     # Set up a new hosts file if the hosts file contains hosts details already or it is an INI file
     if grep -q -e 'hosts:' -e ']' "$CEPH_ANSIBLE_HOSTS" 2>/dev/null; then
```
this is where I think the -s test applies...i.e if the file exists and is not empty, we save it - otherwise it's empty and ready to use. As I said earlier an inventory doesn't have to contain groups [group] blocks, it could just be a list of servers - this is why I'm suggested, if the file exists and contains anything at all...save it, tell the user, and set up a new one for the install.

pcuzner reviewed (changes requested):
After testing, using touch to establish the inventory file creates a problem in the runner-service. If the file exists, it assumes it contains an inventory - if it doesn't exist runner-service seeds it for you. With touch, the inventory file is there - but it's not valid yaml - so the load returns a null inventory that causes 409 errors back to the UI.

TimothyAsirJeyasing committed:
Relocate the hosts file to /usr/share/ceph-ansible

Rename /usr/share/ceph-ansible/hosts if it is already exists
so that the cockpit-ceph-installer will be using a new file.

Create a link file at /usr/share/ansible-runner-service/inventory/hosts
to point /usr/share/ceph-ansible/host file and report an error on
any failures.

Bug: BZ#1879836
https://github.com/red-hat-storage/cockpit-ceph-installer/issues/79
Signed-off-by: Timothy Asir Jeyasingh <tjeyasin@redhat.com>

pcuzner reviewed (changes requested):
This is still problematic. If the file exists from a prior run (ie. it's valid), we rename it and start a new empty inventory...that's not right.

This is the logic we discussed;
<pre>
If the symlink from /usr/share/ansible-runner-service/inventory/hosts is in place
  we do nothing (this implies its a 2nd run, and the hosts file is valid)
else
  if there is a hosts file in /usr/share/ceph-ansible
      rename it (out of the way, so there isn't a hosts file actually in ceph-ansible's dir)
  create the 'hosts' symlink from runner-service to the ceph-ansible dir
</pre>

TimothyAsirJeyasing committed:
Relocate the hosts file to /usr/share/ceph-ansible

Rename /usr/share/ceph-ansible/hosts if it is already exists
and 'hosts' symlink from runner-service is not in place
and create the 'hosts' symlink from runner-service to the ceph-ansible
directory and report any failures.

Bug: BZ#1879836
https://github.com/red-hat-storage/cockpit-ceph-installer/issues/79

Signed-off-by: Timothy Asir Jeyasingh <tjeyasin@redhat.com>

TimothyAsirJeyasing committed:
Relocate the hosts file to /usr/share/ceph-ansible

Rename /usr/share/ceph-ansible/hosts if it is already exists
and 'hosts' symlink from runner-service is not in place
and create the 'hosts' symlink from runner-service to the ceph-ansible
directory and report any failures.

Bug: BZ#1879836
https://github.com/red-hat-storage/cockpit-ceph-installer/issues/79
Signed-off-by: Timothy Asir Jeyasingh <tjeyasin@redhat.com>

pcuzner reviewed (approved):
LGTM - thanks!
